rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,417 @@
+        """
+        sender = transaction.from_address
+        in_event, out_event = self._get_in_out_events(decoded_events, sender)
+        if in_event is None or out_event is None:
+            log.error(
+                f'Failed to post decoding rainbow {transaction.tx_hash.hex()} '
+                f'due to missing in_event or out_event.',
+            )
+            return decoded_events
+
+        out_asset_is_evm_token = out_event.asset.is_evm_token()
+        in_asset_is_evm_token = in_event.asset.is_evm_token()
+
+        if out_event.asset == self.evm_inquirer.native_token and in_asset_is_evm_token:
+            # Try to find if have wrap eth into weth log
+            if (wrap_event := self._find_event_by_log(all_logs, WETH_DEPOSIT_TOPIC)) is None:
+                log.error(
+                    f'Failed to decode fee from swap involving native token to token '
+                    f'for transaction {transaction.tx_hash.hex()} due to missing wrap log.',
+                )
+                return decoded_events
+
+            self._maybe_decode_swap_native_to_token(
+                transaction=transaction,
+                event=out_event,
+                decoded_events=decoded_events,
+                sender=sender,
+                wrap_event=wrap_event,
+            )
+        elif out_asset_is_evm_token and in_event.asset == self.evm_inquirer.native_token:
+            if (unwrap_event := self._find_event_by_log(all_logs, WETH_WITHDRAW_TOPIC)) is None:
+                log.error(
+                    f'Failed to decode fee from swap involving native token to token '
+                    f'for transaction {transaction.tx_hash.hex()} due to missing unwrap log.',
+                )
+                return decoded_events
+
+            self._maybe_decode_fee_from_swap_token_to_native(
+                transaction=transaction,
+                decoded_events=decoded_events,
+                in_event=in_event,
+                out_event=out_event,
+                sender=sender,
+                unwrap_event=unwrap_event,
+            )
+        elif out_asset_is_evm_token and in_asset_is_evm_token:
+            self._maybe_decode_fee_from_swap_token_to_token(
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                sender=sender,
+                out_event=out_event,
+            )
+
+        return decoded_events
+
+    def _maybe_enricher_rainbow_transfer(
+            self,
+            context: EnricherContext,
+    ) -> TransferEnrichmentOutput:
+        """
+        Enricher the event to rainbow event
+        """
+        sender = context.transaction.from_address
+        event = context.event
+        self._enricher_event(event, sender)
+        if event.counterparty == CPT_RAINBOW_SWAPS:
+            return TransferEnrichmentOutput(matched_counterparty=CPT_RAINBOW_SWAPS)
+        return FAILED_ENRICHMENT_OUTPUT
```

Comment:
```suggestion
        """
        Enricher transfer events to rainbow router. The contract doesn't emit events for swaps.
        """
        if event.address != self.router_address:
             return FAILED_ENRICHMENT_OUTPUT
             
        self._enricher_event(event=context.event, context.transaction.from_address)
        if event.counterparty == CPT_RAINBOW_SWAPS:
            return TransferEnrichmentOutput(matched_counterparty=CPT_RAINBOW_SWAPS)
        return FAILED_ENRICHMENT_OUTPUT
```

always exit as soon as possible. You don't need a extra call or read properties if you can exit fast. This is important for enrichers